File: packages/javascript/lib/array/arrCompact.ts
export function arrCompact(array: any[]): any[] {
	if (!Array.isArray(array)) {
		return [];
	}

	const result: any[] = [];

	for (let i = 0, arrayLength = array.length; i < arrayLength; i += 1) {
		const value = array[i];

		// The rejected set is spelled out rather than left to `if (value)`. Dart and Python
		// have no truthiness of their own, so all three implementations have to agree on
		// exactly which values are dropped. Empty arrays and empty objects are truthy in
		// JavaScript and are therefore kept.
		if (
			value === null ||
			value === undefined ||
			value === false ||
			value === 0 ||
			value === '' ||
			(typeof value === 'number' && Number.isNaN(value))
		) {
			continue;
		}

		result.push(value);
	}

	return result;
}

File: packages/python/qsu/array/arrCompact.py
import math


def _isFalsy(value) -> bool:
	# The rejected set is spelled out rather than left to Python's own truthiness, so all
	# three languages agree on exactly which values are dropped. Empty containers are
	# truthy in JavaScript and are therefore kept here as well.
	if value is None or value is False:
		return True

	if isinstance(value, str):
		return value == ''

	if isinstance(value, bool):
		return False

	if isinstance(value, (int, float)):
		if isinstance(value, float) and math.isnan(value):
			return True
		return value == 0

	return False


def arrCompact(array) -> list:
	if not isinstance(array, (list, tuple)):
		return []

	return [value for value in array if not _isFalsy(value)]
